File: packages/i18nify-js/src/modules/core/__tests__/flattenKeys.test.ts
import flattenKeys from '../flattenKeys';

jest.mock('../../../common/errorBoundary', () => ({
  withErrorBoundary: jest.fn((fn) => fn),
}));

describe('flattenKeys', () => {
  it('flattens nested objects with the default delimiter', () => {
    expect(
      flattenKeys({
        user: {
          name: 'Ada',
          address: {
            city: 'London',
          },
        },
      }),
    ).toEqual({
      'user.name': 'Ada',
      'user.address.city': 'London',
    });
  });

  it('preserves arrays and null values as leaf nodes', () => {
    expect(
      flattenKeys({
        user: {
          aliases: ['Ada', 'Lovelace'],
          preferences: null,
        },
      }),
    ).toEqual({
      'user.aliases': ['Ada', 'Lovelace'],
      'user.preferences': null,
    });
  });

  it('supports a custom delimiter', () => {
    expect(
      flattenKeys(
        {
          user: {
            profile: {
              city: 'Mumbai',
            },
          },
        },
        { delimiter: '/' },
      ),
    ).toEqual({
      'user/profile/city': 'Mumbai',
    });
  });

  it('throws when the input is not a plain object', () => {
    expect(() => flattenKeys([] as unknown as Record<string, unknown>)).toThrow(
      'flattenKeys: input must be a plain object.',
    );
  });

  it('throws when the delimiter is empty', () => {
    expect(() =>
      flattenKeys({ user: { name: 'Ada' } }, { delimiter: '' }),
    ).toThrow('flattenKeys: delimiter must be a non-empty string.');
  });
});

File: packages/i18nify-go/modules/core/flatten_keys_test.go
package core

import (
	"reflect"
	"testing"
)

func TestFlattenKeys(t *testing.T) {
	tests := []struct {
		name    string
		input   map[string]any
		opts    *FlattenOptions
		want    map[string]any
		wantErr bool
	}{
		{
			name: "flattens nested maps with default delimiter",
			input: map[string]any{
				"user": map[string]any{
					"name": "Ada",
					"address": map[string]any{
						"city": "London",
					},
				},
			},
			want: map[string]any{
				"user.name":         "Ada",
				"user.address.city": "London",
			},
		},
		{
			name: "preserves arrays as leaf values",
			input: map[string]any{
				"user": map[string]any{
					"aliases": []string{"Ada", "Lovelace"},
				},
			},
			want: map[string]any{
				"user.aliases": []string{"Ada", "Lovelace"},
			},
		},
		{
			name: "supports custom delimiter",
			input: map[string]any{
				"user": map[string]any{
					"profile": map[string]any{
						"city": "Mumbai",
					},
				},
			},
			opts: &FlattenOptions{Delimiter: "/"},
			want: map[string]any{
				"user/profile/city": "Mumbai",
			},
		},
		{
			name:    "returns error for nil input",
			wantErr: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := FlattenKeys(tt.input, tt.opts)
			if tt.wantErr {
				if err == nil {
					t.Fatal("expected an error but got nil")
				}
				return
			}
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			if !reflect.DeepEqual(got, tt.want) {
				t.Fatalf("FlattenKeys() = %#v, want %#v", got, tt.want)
			}
		})
	}
}
